### ISSUE TITLE
Support models having composite primary keys

### ISSUE THREAD

**Issue by kakra** [mixonic/ranked-model#37]
I'm trying to use a table from a legacy DB which has no singular primary key. Instead I used the composite-primary-keys gem to use this table in Rails. When trying to use ranked-model, the SQL breaks (do not look too much at the ugly field names, that will be changed later when the legacy applications using the DB become obsolete):

```
irb(main):050:0> pos = Article.find(34129).topic_article_links.find_by_topicid(348)
  Article Load (0.6ms)  SELECT `topicarticles`.* FROM `topicarticles` WHERE `topicarticles`.`articleid` = 34129 LIMIT 1
  TopicArticleLink Load (0.5ms)  SELECT `group_topic_article`.* FROM `group_topic_article` WHERE `group_topic_article`.`articleid` = 34129 AND `group_topic_article`.`topicid` = 348 LIMIT 1
=> #<TopicArticleLink topicid: 348, articleid: 34129, status: "AKTIV", article_prio: 0.0>
irb(main):051:0> pos.article_order_position = 0
=> 0
irb(main):052:0> pos.save
   (0.5ms)  BEGIN
  TopicArticleLink Load (11.8ms)  SELECT `group_topic_article`.`id`, `group_topic_article`.`article_prio` FROM `group_topic_article` WHERE `group_topic_article`.`topic_id` IS NULL AND (`group_topic_article`.`id` != 348, 34129) ORDER BY `group_topic_article`.`article_prio` ASC LIMIT 1
   (0.1ms)  ROLLBACK
ActiveRecord::StatementInvalid: Mysql2::Error: Unknown column 'group_topic_article.id' in 'field list': SELECT  `group_topic_article`.`id`, `group_topic_article`.`article_prio` FROM `group_topic_article`  WHERE `group_topic_article`.`topic_id` IS NULL AND (`group_topic_article`.`id` != 348, 34129) ORDER BY `group_topic_article`.`article_prio` ASC LIMIT 1
```

`TopicArticleLink` is the join-model between `Article` and `Topic`. The simplified models look like this:

```
class TopicArticleLink < ActiveRecord::Base
  self.table_name = "group_topic_article"
  self.primary_keys = :topicid, :articleid

  belongs_to :topic, :foreign_key => "topicid"
  belongs_to :article, :foreign_key => "articleid"

  include RankedModel
  ranks :article_order, :column => :article_prio, :with_same => :topicid
end

class Article < ActiveRecord::Base
  self.table_name = "topicarticles"
  self.primary_key = "articleid"

  has_many :topic_article_links, :foreign_key => "articleid", :dependent => :destroy
  has_many :topics, :through => :topic_article_links
end

class Topic < ActiveRecord::Base
  self.primary_key = "topicid"
  has_many :topic_article_links, :foreign_key => "topicid", :dependent => :delete_all
  has_many :articles, :through => :topic_article_links
end
```

**Comment by phlegx** [comment]
Hi!

I have write this hook to test ranked-model on my app.

```ruby
# config/initializers/ranked_model.rb

module RankedModel
  class Ranker
    class Mapper
      def update_rank! value
        # Bypass callbacks
        #
        instance_class.
          where(primary_key_query).  # Call added method.
          update_all(ranker.column => value)
      end

      private

      def finder(order = :asc)
        @finder ||= {}
        @finder[order] ||= begin
          _finder = instance_class
          columns = [primary_key, ranker.column].flatten  # Call added method.

          if ranker.scope
            _finder = _finder.send ranker.scope
          end

          case ranker.with_same
          when Symbol
            columns << ranker.with_same
            _finder = _finder.where \
              ranker.with_same => instance.attributes[ranker.with_same.to_s]
          when Array
            ranker.with_same.each do |column|
              columns << column
              _finder = _finder.where column => instance.attributes[column.to_s]
            end
          end

          unless new_record?
            _finder = _finder.where.not primary_key_query # Call added method.
          end

          _finder.reorder(ranker.column.to_sym => order).select(columns)
        end
      end

      def current_order
        @current_order ||= begin
          finder.unscope(where: primary_key).collect { |ordered_instance|
            RankedModel::Ranker::Mapper.new ranker, ordered_instance
          }
        end
      end

      # Added this method to get the right primary_key.
      def primary_key
        if instance_class.primary_key.is_a?(Array)
          instance_class.primary_key.map(&:to_sym)
        else
          instance_class.primary_key.to_sym
        end
      end

      # Added this method to get the right primary_key query.
      def primary_key_query
        if instance_class.primary_key.is_a?(Array)
          instance_class.cpk_id_predicate(instance_class.arel_table, # This method comes from gem composite_primary_keys.
                                          instance_class.primary_key,
                                          instance.id)
        else
          { instance_class.primary_key.to_sym => instance.id }
        end
      end
    end
  end
end
```

What do you think? I use gem `composite_primary_keys`.

**Comment by kakra** [comment]
I abandoned using composite primary keys and converted my models to single primary keys with an additional unique constraint of the columns previously used as composite primary key. This creates much less headaches.

This issue is now 7 years old, and I've moved the project related to this forward by eliminating all composite primary keys. Previously, those only existed due to dependency on legacy components which have since been replaced by native Rails components, and now this is no longer an issue for me.

So from my POV, this can be closed as abandoned.

I'd like to suggest creating a new issue report for your case - so I can close this. We can put a link to this one in the new report so everyone who stumbles across it will find your suggestion. But I'd rather close this one, I'm no longer interested in resolving this, neither seems the author of `ranked-models`. So I'm just following upstream.
